plugins/storage/volume/linstor/src/main/java/org/apache/cloudstack/storage/datastore/driver/LinstorPrimaryDataStoreDriverImpl.java

```diff
@@ -0,0 +1,720 @@
+                    s_logger.error(errMsg);
+            }
+        } catch (Exception ex)
+        {
+            errMsg = ex.getMessage();
+
+            s_logger.error("createAsync: " + errMsg);
+            if (callback == null)
+            {
+                throw ex;
+            }
+        }
+
+        if (callback != null)
+        {
+            CreateCmdResult result = new CreateCmdResult(devPath, new Answer(null, errMsg == null, errMsg));
+            result.setResult(errMsg);
+            callback.complete(result);
+        }
+    }
+
+    @Override
+    public void revertSnapshot(
+        SnapshotInfo snapshot,
+        SnapshotInfo snapshotOnPrimaryStore,
+        AsyncCompletionCallback<CommandResult> callback)
+    {
+        s_logger.debug("Linstor: revertSnapshot");
+        final VolumeInfo volumeInfo = snapshot.getBaseVolume();
+        VolumeVO volumeVO = _volumeDao.findById(volumeInfo.getId());
+        if (volumeVO == null || volumeVO.getRemoved() != null) {
+            CommandResult commandResult = new CommandResult();
+            commandResult.setResult("The volume that the snapshot belongs to no longer exists.");
+            callback.complete(commandResult);
+            return;
+        }
+
+        String resultMsg;
+        try {
+            final StoragePool pool = (StoragePool) snapshot.getDataStore();
+            final String rscName = LinstorUtil.RSC_PREFIX + volumeInfo.getUuid();
+            final String snapName = LinstorUtil.RSC_PREFIX + snapshot.getUuid();
+            final DevelopersApi linstorApi = LinstorUtil.getLinstorAPI(pool.getHostAddress());
+
+            ApiCallRcList answers = linstorApi.resourceSnapshotRollback(rscName, snapName);
+            resultMsg = checkLinstorAnswers(answers);
+        } catch (ApiException apiEx) {
+            s_logger.error("Linstor: ApiEx - " + apiEx.getMessage());
+            resultMsg = apiEx.getBestMessage();
+        }
+
+        if (callback != null)
+        {
+            CommandResult result = new CommandResult();
+            result.setResult(resultMsg);
+            callback.complete(result);
+        }
+    }
+
+    @Override
+    public boolean canCopy(DataObject srcData, DataObject destData)
+    {
+        return false;
+    }
+
+    @Override
+    public void copyAsync(DataObject srcData, DataObject destData, AsyncCompletionCallback<CopyCommandResult> callback)
+    {
+        // as long as canCopy is false, this isn't called
+        s_logger.debug("Linstor: copyAsync with srcdata: " + srcData.getUuid());
+    }
+
+    @Override
+    public void copyAsync(DataObject srcData, DataObject destData, Host destHost, AsyncCompletionCallback<CopyCommandResult> callback)
+    {
+        // as long as canCopy is false, this isn't called
+        s_logger.debug("Linstor: copyAsync with srcdata: " + srcData.getUuid());
+    }
+
+    @Override
+    public void resize(DataObject data, AsyncCompletionCallback<CreateCmdResult> callback)
```

Comment:
@rp- I'm not sure if I'm right, but when you do a live resize, you're resizing the volume only on Linstor, but the guest/hypervisor is not informed. That's because the Linstor primary storage is not managed, and you have to notify qemu from here. Please excuse me if I'm wrong

Reply:
I tried resizing, indeed online resizing isn't noticed by the guest. But if i power down and start the guest again. it has the new size available. I looked into other storage drivers, but didn't really see were they would notify the virtualization platform of the size change?

Reply:
Do we do this for other storage types? I'm not aware we support online disk resize for other storage platforms, but I may be wrong.

Reply:
@rp-, most of the plugins in CloudStack are for managed storage. In your case I think it will be useful for you to check the default driver implementation - `CloudStackPrimaryDataStoreDriverImpl.resize()` and maybe you have to modify `LibvirtResizeVolumeCommandWrapper` or to add a custom `CommandWrapper`.

Reply:
@andrijapanicsb, we at StorPool support live resize. It is supported for NFS/Ceph also

Reply:
I added the notify call, but it didn't make a difference for my alpine guest.

Reply:
@rp- is there any output when 
`/usr/share/cloudstack-common/scripts/storage/qcow2/resizevolume.sh -s
 newSize -c currentSize -p volumePath -t NOTIFYONLY -r shrink -v vmName` is executed?

Reply:
just got the following:
```
Sep 08 08:52:51 cloudstack-comp2.linbit java[6215]: INFO  [resource.wrapper.LibvirtResizeVolumeCommandWrapper] (agentRequest-Handler-1:) (logid:08a8fb35) Invoking resize script to handle type NOTIFYONLY
Sep 08 08:52:51 cloudstack-comp2.linbit java[6215]: DEBUG [resource.wrapper.LibvirtResizeVolumeCommandWrapper] (agentRequest-Handler-1:) (logid:08a8fb35) Executing: /usr/share/cloudstack-common/scripts/storage
/qcow2/resizevolume.sh -s 10737418240 -c 9663676416 -p /dev/drbd1001 -t NOTIFYONLY -r false -v i-2-253-VM
Sep 08 08:52:51 cloudstack-comp2.linbit java[6215]: DEBUG [resource.wrapper.LibvirtResizeVolumeCommandWrapper] (agentRequest-Handler-1:) (logid:08a8fb35) Executing while with timeout : 7200000
Sep 08 08:52:51 cloudstack-comp2.linbit java[6215]: DEBUG [resource.wrapper.LibvirtResizeVolumeCommandWrapper] (agentRequest-Handler-1:) (logid:08a8fb35) Execution is successful.
```

Reply:
Hmm, this tells that the notification of the guest was successful. Is there any change when you run:
`virsh domblkinfo --device /dev/drbd1001 --domain i-2-253-VM --human`
or
`lsblk` from the guest?

Reply:
```
[root@cloudstack-comp2 lib]# virsh domblkinfo --device /dev/drbd1001
 --domain i-2-253-VM --human
Capacity:       10.002 GiB
Allocation:     1.102 GiB
Physical:       10.002 GiB
```

lsblk in guest still tells me the old size